### ISSUE TITLE
Casts from `const void *` in a constant expression sometimes produce the wrong diagnostic

### ISSUE THREAD

**Issue by tbaederr** [llvm/llvm-project#62467]
See: https://godbolt.org/z/M5PdG83Ks

```c++
constexpr bool f(const int *p) {
  return (int*)p;
}

static_assert(f((const void *)nullptr));
```

This code produces the following diagnostic:

```console 
<source>:6:15: error: no matching function for call to 'f'
static_assert(f((const void *)nullptr));
              ^
<source>:2:16: note: candidate function not viable: cannot convert argument of incomplete type 'const void *' to 'const int *' for 1st argument
constexpr bool f(const int *p) {
               ^
1 error generated.
```

Which... makes no sense? "Incomplete type void *"?
GCC manages to produce the diagnostic I expected to see.

**Comment by shafik** [comment]
I believe clang is correct here, see [basic.type.general p5](https://eel.is/c++draft/basic.types.general#5):

> ... Incompletely-defined object types and cv void are incomplete types ([basic.fundamental])

I think it is a little awkward perhaps CC @AaronBallman

**Comment by AaronBallman** [comment]
I think the diagnostic is wrong -- `const void *` is not an incomplete type, `void` is. This is a garden variety type conversion issue. If you change `const void *` to `const char *` you get the much more comprehensible note `note: candidate function not viable: no known conversion from 'const char *' to 'const int *' for 1st argument` which is about what I'd expect in the `const void *` case as well.

**Comment by hazohelet** [comment]
I looked into the code and now think this is a wording issue.
There is a policy to have only one note for each overload candidate not to make too many notes:
https://github.com/llvm/llvm-project/blob/6b4bb103db37ec11507bf3f3fa26731bd16501a5/clang/lib/Sema/SemaOverload.cpp#L11461-L11466
So, I propose to change the wording introduced in https://github.com/llvm/llvm-project/commit/6d174646dd1a84127128f98cbef0203f23df6f12 (and also remove `void*` from this specialized case) to say that `TempFromTy` is the incomplete type, but I find it hard to think of a proper message here.

Do you have any idea about the proper wording or another approach here? @AaronBallman @cjdb

**Comment by AaronBallman** [comment]
I don't think I would reword the diagnostic (much); instead, I would skip emitting it there. I'd change the predicate to:
```
if (TempFromTy->isIncompleteType() && TempFromTy->isRecordType()) {
```
because I think that's the only case that was considered for the changes adding this code. There's basically no chance that `void *` is causing the overload issue due to its incompleteness, but this also covers weird edge cases like `static_assert(f((enum DOES_NOT_EXIST *)nullptr));` which currently talk about not being able to convert the incomplete type. Using a predicate that only considers incomplete structure and union types seems like the right call to me. WDYT @rjmccall?

I would not be sad if the diagnostic was reworded to be more precise: `... cannot convert argument of %select{pointer|reference}N to incomplete type ...` but that might just be me being pedantic with diagnostic wording. @cjdb (or others) may have different opinions.
